File: cra-pom-v2/src/core/v3/geometry/homology.ts
/**
 * Homological Analysis for Geometric Cognition
 *
 * Extracts topological signals (Betti numbers) from geometric states.
 * This enables detection of structural patterns that are invariant under
 * continuous deformation.
 *
 * BETTI NUMBERS:
 * - b₀: Connected components (clusters of related concepts)
 * - b₁: 1-dimensional holes (loops/cycles in relationships)
 * - b₂: 2-dimensional voids (surfaces enclosing empty space)
 * - b₃: 3-dimensional voids (volumes in 4D)
 *
 * COGNITIVE SIGNIFICANCE:
 * - b₀ change = new concept cluster formed/merged
 * - b₁ increase = cyclic relationship detected (e.g., circle of fifths!)
 * - b₂ void = conceptual gap or contradiction
 * - Persistence = stability of topological features across scales
 *
 * IMPLEMENTATION:
 * Uses simplicial complex approximation via Rips complex construction.
 * At each radius threshold, we build simplices from nearby points and
 * compute homology groups.
 *
 * REFERENCES:
 * - Edelsbrunner & Harer, "Computational Topology"
 * - https://en.wikipedia.org/wiki/Persistent_homology
 */

import type { Vector4D } from '../music/music-geometry-domain';
import { distance4D } from '../music/polytopes';

// ============================================================================
// Types
// ============================================================================

export interface BettiNumbers {
  b0: number;  // Connected components
  b1: number;  // 1D holes (loops)
  b2: number;  // 2D voids (cavities)
  b3: number;  // 3D voids (in 4D)
}

export interface PersistenceInterval {
  birth: number;    // Threshold where feature appears
  death: number;    // Threshold where feature disappears (Infinity = never)
  dimension: number;  // Which Betti number (0, 1, 2, or 3)
}

export interface PersistenceDiagram {
  intervals: PersistenceInterval[];
  maxRadius: number;
}

export type TopologicalEventType =
  | 'cluster_formed'      // b₀ increased
  | 'cluster_merged'      // b₀ decreased
  | 'loop_formed'         // b₁ increased
  | 'loop_filled'         // b₁ decreased
  | 'void_formed'         // b₂ increased
  | 'void_filled'         // b₂ decreased
  | 'hypervoid_formed'    // b₃ increased
  | 'hypervoid_filled';   // b₃ decreased

export interface TopologicalEvent {
  type: TopologicalEventType;
  dimension: number;
  delta: number;
  threshold: number;
}

export interface SimplexData {
  vertices: number[];   // Vertex indices
  dimension: number;    // 0 = point, 1 = edge, 2 = triangle, 3 = tetrahedron
}

// ============================================================================
// Distance Matrix
// ============================================================================

/**
 * Compute pairwise distance matrix for points
 */
export function computeDistanceMatrix(points: Vector4D[]): number[][] {
  const n = points.length;
  const matrix: number[][] = [];

  for (let i = 0; i < n; i++) {
    matrix[i] = [];
    for (let j = 0; j < n; j++) {
      if (i === j) {
        matrix[i][j] = 0;
      } else if (j < i) {
        matrix[i][j] = matrix[j][i];
      } else {
        matrix[i][j] = distance4D(points[i], points[j]);
      }
    }
  }

  return matrix;
}

// ============================================================================
// Rips Complex Construction
// ============================================================================

/**
 * Build Rips complex at given threshold
 * A simplex is included if all pairwise distances are ≤ threshold
 */
export function buildRipsComplex(
  distanceMatrix: number[][],
  threshold: number,
  maxDimension: number = 3
): SimplexData[] {
  const n = distanceMatrix.length;
  const simplices: SimplexData[] = [];

  // 0-simplices (vertices) - all points
  for (let i = 0; i < n; i++) {
    simplices.push({ vertices: [i], dimension: 0 });
  }

  // 1-simplices (edges) - pairs within threshold
  const edges: [number, number][] = [];
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if (distanceMatrix[i][j] <= threshold) {
        simplices.push({ vertices: [i, j], dimension: 1 });
        edges.push([i, j]);
      }
    }
  }

  if (maxDimension < 2) return simplices;

  // 2-simplices (triangles) - check all triples
  const triangles: number[][] = [];
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if (distanceMatrix[i][j] > threshold) continue;
      for (let k = j + 1; k < n; k++) {
        if (distanceMatrix[i][k] <= threshold &&
            distanceMatrix[j][k] <= threshold) {
          simplices.push({ vertices: [i, j, k], dimension: 2 });
          triangles.push([i, j, k]);
        }
      }
    }
  }

  if (maxDimension < 3) return simplices;

  // 3-simplices (tetrahedra) - check all quadruples
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if (distanceMatrix[i][j] > threshold) continue;
      for (let k = j + 1; k < n; k++) {
        if (distanceMatrix[i][k] > threshold ||
            distanceMatrix[j][k] > threshold) continue;
        for (let l = k + 1; l < n; l++) {
          if (distanceMatrix[i][l] <= threshold &&
              distanceMatrix[j][l] <= threshold &&
              distanceMatrix[k][l] <= threshold) {
            simplices.push({ vertices: [i, j, k, l], dimension: 3 });
          }
        }
      }
    }
  }

  return simplices;
}

// ============================================================================
// Homology Computation (Simplified)
// ============================================================================

/**
 * Compute connected components (b₀) using union-find
 */
function computeB0(simplices: SimplexData[], n: number): number {
  // Union-find for connected components
  const parent = new Array(n).fill(-1);

  function find(x: number): number {
    if (parent[x] < 0) return x;
    parent[x] = find(parent[x]);
    return parent[x];
  }

  function union(x: number, y: number): void {
    const px = find(x);
    const py = find(y);
    if (px !== py) {
      // Union by rank
      if (parent[px] < parent[py]) {
        parent[px] += parent[py];
        parent[py] = px;
      } else {
        parent[py] += parent[px];
        parent[px] = py;
      }
    }
  }

  // Process all edges
  for (const simplex of simplices) {
    if (simplex.dimension === 1) {
      union(simplex.vertices[0], simplex.vertices[1]);
    }
  }

  // Count roots (negative parent values)
  let components = 0;
  for (let i = 0; i < n; i++) {
    if (parent[i] < 0) components++;
  }

  return components;
}

/**
 * Compute b₁ (loops) using Euler characteristic approximation
 * χ = V - E + F - T + ...
 * For simplicial complex: b₀ - b₁ + b₂ - b₃ = χ
 */
function computeBettiApprox(simplices: SimplexData[], n: number): BettiNumbers {
  // Count simplices by dimension
  const counts = [0, 0, 0, 0]; // vertices, edges, faces, tetrahedra
  for (const simplex of simplices) {
    if (simplex.dimension <= 3) {
      counts[simplex.dimension]++;
    }
  }

  const b0 = computeB0(simplices, n);

  // Use Euler characteristic relationship for approximation
  // This is a simplification; true computation requires boundary operator
  const euler = counts[0] - counts[1] + counts[2] - counts[3];

  // Approximate b₁ from Euler characteristic: χ ≈ b₀ - b₁ + b₂ - b₃
  // For typical point clouds, b₂ and b₃ are often small or zero initially
  // Rough estimate: b₁ ≈ b₀ - χ (assuming b₂, b₃ small)
  const b1Est = Math.max(0, b0 - euler);

  // b₂ approximation based on void detection
  // A void requires a closed surface of triangles with no tetrahedra inside
  // Simplified: b₂ ≈ triangles that don't belong to any tetrahedron / factor
  const b2Est = Math.max(0, Math.floor(counts[2] / 20) - Math.floor(counts[3] / 2));

  // b₃ for 4D voids (very rare in typical data)
  const b3Est = 0;

  return {
    b0,
    b1: b1Est,
    b2: b2Est,
    b3: b3Est,
  };
}

/**
 * Compute Betti numbers from point cloud at given threshold
 */
export function computeBetti(points: Vector4D[], threshold: number): BettiNumbers {
  if (points.length === 0) {
    return { b0: 0, b1: 0, b2: 0, b3: 0 };
  }

  const distMatrix = computeDistanceMatrix(points);
  const simplices = buildRipsComplex(distMatrix, threshold, 3);
  return computeBettiApprox(simplices, points.length);
}

// ============================================================================
// Persistent Homology
// ============================================================================

/**
 * Compute persistence diagram by tracking Betti numbers across thresholds
 */
export function computePersistence(
  points: Vector4D[],
  numThresholds: number = 20,
  maxRadius?: number
): PersistenceDiagram {
  if (points.length === 0) {
    return { intervals: [], maxRadius: 0 };
  }

  // Determine max radius from data if not provided
  const distMatrix = computeDistanceMatrix(points);
  let computedMaxRadius = maxRadius ?? 0;

  if (!maxRadius) {
    for (let i = 0; i < points.length; i++) {
      for (let j = i + 1; j < points.length; j++) {
        if (distMatrix[i][j] > computedMaxRadius) {
          computedMaxRadius = distMatrix[i][j];
        }
      }
    }
  }

  // Sample thresholds
  const thresholds: number[] = [];
  for (let i = 0; i <= numThresholds; i++) {
    thresholds.push((i / numThresholds) * computedMaxRadius);
  }

  // Track Betti numbers at each threshold
  const history: { threshold: number; betti: BettiNumbers }[] = [];
  for (const threshold of thresholds) {
    const simplices = buildRipsComplex(distMatrix, threshold, 3);
    const betti = computeBettiApprox(simplices, points.length);
    history.push({ threshold, betti });
  }

  // Extract persistence intervals
  const intervals: PersistenceInterval[] = [];

  // Track birth/death of features
  for (let dim = 0; dim <= 3; dim++) {
    let prevCount = 0;
    const births: number[] = [];

    for (const { threshold, betti } of history) {
      const bettiKey = `b${dim}` as keyof BettiNumbers;
      const count = betti[bettiKey];

      // New features born
      while (births.length < count) {
        births.push(threshold);
      }

      // Features died
      while (births.length > count) {
        const birth = births.pop()!;
        intervals.push({
          birth,
          death: threshold,
          dimension: dim,
        });
      }

      prevCount = count;
    }

    // Remaining features persist to infinity
    for (const birth of births) {
      intervals.push({
        birth,
        death: Infinity,
        dimension: dim,
      });
    }
  }

  return {
    intervals,
    maxRadius: computedMaxRadius,
  };
}

/**
 * Get persistence (death - birth) for an interval
 */
export function getIntervalPersistence(interval: PersistenceInterval): number {
  if (interval.death === Infinity) return Infinity;
  return interval.death - interval.birth;
}

/**
 * Filter persistence diagram to significant features
 */
export function filterPersistence(
  diagram: PersistenceDiagram,
  minPersistence: number
): PersistenceInterval[] {
  return diagram.intervals.filter(
    interval => getIntervalPersistence(interval) >= minPersistence
  );
}

// ============================================================================
// Topological Event Detection
// ============================================================================

/**
 * Detect topological changes between two states
 */
export function detectTransition(
  before: BettiNumbers,
  after: BettiNumbers,
  threshold: number = 0
): TopologicalEvent[] {
  const events: TopologicalEvent[] = [];

  // Check each dimension
  const checks: { dim: number; key: keyof BettiNumbers; incType: TopologicalEventType; decType: TopologicalEventType }[] = [
    { dim: 0, key: 'b0', incType: 'cluster_formed', decType: 'cluster_merged' },
    { dim: 1, key: 'b1', incType: 'loop_formed', decType: 'loop_filled' },
    { dim: 2, key: 'b2', incType: 'void_formed', decType: 'void_filled' },
    { dim: 3, key: 'b3', incType: 'hypervoid_formed', decType: 'hypervoid_filled' },
  ];

  for (const { dim, key, incType, decType } of checks) {
    const delta = after[key] - before[key];
    if (delta > 0) {
      events.push({
        type: incType,
        dimension: dim,
        delta,
        threshold,
      });
    } else if (delta < 0) {
      events.push({
        type: decType,
        dimension: dim,
        delta: Math.abs(delta),
        threshold,
      });
    }
  }

  return events;
}

// ============================================================================
// HomologyAnalyzer Class
// ============================================================================

export class HomologyAnalyzer {
  private points: Vector4D[];
  private distanceMatrix: number[][] | null = null;
  private cachedBetti: Map<number, BettiNumbers> = new Map();

  constructor(points: Vector4D[] = []) {
    this.points = points;
  }

  /**
   * Set/update the point cloud
   */
  setPoints(points: Vector4D[]): void {
    this.points = points;
    this.distanceMatrix = null;
    this.cachedBetti.clear();
  }

  /**
   * Get distance matrix (cached)
   */
  getDistanceMatrix(): number[][] {
    if (!this.distanceMatrix) {
      this.distanceMatrix = computeDistanceMatrix(this.points);
    }
    return this.distanceMatrix;
  }

  /**
   * Compute Betti numbers at threshold (cached)
   */
  computeBetti(threshold: number): BettiNumbers {
    // Round threshold for caching
    const key = Math.round(threshold * 1000) / 1000;
    if (this.cachedBetti.has(key)) {
      return this.cachedBetti.get(key)!;
    }

    const distMatrix = this.getDistanceMatrix();
    const simplices = buildRipsComplex(distMatrix, threshold, 3);
    const betti = computeBettiApprox(simplices, this.points.length);

    this.cachedBetti.set(key, betti);
    return betti;
  }

  /**
   * Compute persistence diagram
   */
  computePersistence(numThresholds: number = 20): PersistenceDiagram {
    return computePersistence(this.points, numThresholds);
  }

  /**
   * Find optimal threshold where b₁ first appears (loop detection)
   */
  findFirstLoop(): number | null {
    const persistence = this.computePersistence(50);
    const loopIntervals = persistence.intervals.filter(i => i.dimension === 1);

    if (loopIntervals.length === 0) return null;

    // Find minimum birth time
    let minBirth = Infinity;
    for (const interval of loopIntervals) {
      if (interval.birth < minBirth) {
        minBirth = interval.birth;
      }
    }

    return minBirth === Infinity ? null : minBirth;
  }

  /**
   * Detect circle of fifths structure
   * Should show b₁ = 1 at appropriate threshold
   */
  detectCyclicStructure(): { threshold: number; loopCount: number } | null {
    const persistence = this.computePersistence(30);

    // Find persistent loops (long-lived b₁ features)
    const significantLoops = filterPersistence(persistence, persistence.maxRadius * 0.1)
      .filter(i => i.dimension === 1);

    if (significantLoops.length === 0) return null;

    // Find threshold where we have the most loops
    let bestThreshold = significantLoops[0].birth;
    let maxLoops = 0;

    for (let t = 0; t <= persistence.maxRadius; t += persistence.maxRadius / 30) {
      const betti = this.computeBetti(t);
      if (betti.b1 > maxLoops) {
        maxLoops = betti.b1;
        bestThreshold = t;
      }
    }

    return { threshold: bestThreshold, loopCount: maxLoops };
  }

  /**
   * Get summary of topological features
   */
  getSummary(threshold: number): {
    betti: BettiNumbers;
    description: string[];
  } {
    const betti = this.computeBetti(threshold);
    const description: string[] = [];

    if (betti.b0 === 1) {
      description.push('All points connected (single cluster)');
    } else if (betti.b0 > 1) {
      description.push(`${betti.b0} separate clusters`);
    }

    if (betti.b1 > 0) {
      description.push(`${betti.b1} loop(s)/cycle(s) detected`);
    }

    if (betti.b2 > 0) {
      description.push(`${betti.b2} void(s)/cavity(ies) detected`);
    }

    if (betti.b3 > 0) {
      description.push(`${betti.b3} 4D hypervoid(s) detected`);
    }

    if (description.length === 0) {
      description.push('No topological features at this threshold');
    }

    return { betti, description };
  }

  /**
   * Get number of points
   */
  get pointCount(): number {
    return this.points.length;
  }
}

// ============================================================================
// Factory Functions
// ============================================================================

/**
 * Create analyzer for polytope vertices
 */
export function createPolytopeAnalyzer(vertices: Vector4D[]): HomologyAnalyzer {
  return new HomologyAnalyzer(vertices);
}

/**
 * Quick Betti computation for a set of points
 */
export function quickBetti(points: Vector4D[], threshold: number): BettiNumbers {
  return computeBetti(points, threshold);
}

// ============================================================================
// Exports
// ============================================================================

export const HomologyModule = {
  // Core computation
  computeBetti,
  computeDistanceMatrix,
  buildRipsComplex,

  // Persistence
  computePersistence,
  getIntervalPersistence,
  filterPersistence,

  // Event detection
  detectTransition,

  // Class
  HomologyAnalyzer,

  // Factory
  createPolytopeAnalyzer,
  quickBetti,
};

export default HomologyModule;

File: rust-engine/src/analysis/homology.rs
//! Homological Analysis - Betti Numbers and Persistent Homology
//!
//! Computes topological invariants of the point cloud/image to detect
//! structural features like clusters, loops, and voids.

use crate::geometry::Vec4;
use serde::{Serialize, Deserialize};

/// Betti numbers characterizing the topology
#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize)]
pub struct BettiNumbers {
    /// β₀: Number of connected components (clusters)
    pub b0: usize,
    /// β₁: Number of 1-dimensional holes (loops)
    pub b1: usize,
    /// β₂: Number of 2-dimensional voids (cavities)
    pub b2: usize,
    /// β₃: Number of 3-dimensional voids (in 4D)
    pub b3: usize,
}

impl BettiNumbers {
    /// Total topological complexity
    pub fn total(&self) -> usize {
        self.b0 + self.b1 + self.b2 + self.b3
    }

    /// Euler characteristic (alternating sum)
    pub fn euler_characteristic(&self) -> i64 {
        self.b0 as i64 - self.b1 as i64 + self.b2 as i64 - self.b3 as i64
    }

    /// Is this topologically trivial (single connected component, no holes)?
    pub fn is_trivial(&self) -> bool {
        self.b0 == 1 && self.b1 == 0 && self.b2 == 0 && self.b3 == 0
    }
}

/// A topological signal detected in the data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopologicalSignal {
    /// Signal type
    pub signal_type: SignalType,
    /// Persistence (birth - death time)
    pub persistence: f64,
    /// Location (approximate center)
    pub location: [f64; 4],
    /// Scale at which it appears
    pub scale: f64,
}

/// Types of topological signals
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum SignalType {
    /// New cluster formed
    ClusterBirth,
    /// Clusters merged
    ClusterMerge,
    /// Loop formed
    LoopFormation,
    /// Loop closed/filled
    LoopFilled,
    /// Void formed
    VoidFormation,
    /// Void collapsed
    VoidCollapse,
    /// Symmetry detected
    SymmetryDetected,
}

/// Homological analyzer
pub struct HomologyAnalyzer {
    /// Distance threshold for connectivity
    pub distance_threshold: f64,
    /// Minimum persistence to report
    pub min_persistence: f64,
    /// Last computed Betti numbers
    last_betti: BettiNumbers,
}

impl HomologyAnalyzer {
    pub fn new() -> Self {
        Self {
            distance_threshold: 1.5,
            min_persistence: 0.1,
            last_betti: BettiNumbers::default(),
        }
    }

    /// Compute Betti numbers for a set of 4D points
    pub fn compute_betti(&mut self, points: &[Vec4]) -> BettiNumbers {
        if points.is_empty() {
            return BettiNumbers::default();
        }

        // Simplified Betti number computation using connectivity analysis
        // In a full implementation, we'd use persistent homology (e.g., Ripser)

        // β₀: Count connected components using union-find
        let b0 = self.count_components(points);

        // β₁: Estimate loops from graph cycles
        let b1 = self.estimate_loops(points);

        // β₂: Estimate voids (simplified)
        let b2 = self.estimate_voids(points);

        // β₃: 4D voids (very simplified)
        let b3 = 0;

        let betti = BettiNumbers { b0, b1, b2, b3 };
        self.last_betti = betti;
        betti
    }

    /// Count connected components using union-find
    fn count_components(&self, points: &[Vec4]) -> usize {
        let n = points.len();
        if n == 0 {
            return 0;
        }

        // Simple union-find
        let mut parent: Vec<usize> = (0..n).collect();

        fn find(parent: &mut [usize], i: usize) -> usize {
            if parent[i] != i {
                parent[i] = find(parent, parent[i]);
            }
            parent[i]
        }

        fn union(parent: &mut [usize], i: usize, j: usize) {
            let pi = find(parent, i);
            let pj = find(parent, j);
            if pi != pj {
                parent[pi] = pj;
            }
        }

        // Connect nearby points
        let threshold_sq = self.distance_threshold * self.distance_threshold;
        for i in 0..n {
            for j in (i + 1)..n {
                if points[i].distance_squared(points[j]) < threshold_sq {
                    union(&mut parent, i, j);
                }
            }
        }

        // Count unique roots
        let mut roots = std::collections::HashSet::new();
        for i in 0..n {
            roots.insert(find(&mut parent, i));
        }

        roots.len()
    }

    /// Estimate number of loops from graph structure
    fn estimate_loops(&self, points: &[Vec4]) -> usize {
        let n = points.len();
        if n < 3 {
            return 0;
        }

        // Build adjacency graph
        let threshold_sq = self.distance_threshold * self.distance_threshold;
        let mut edge_count = 0;

        for i in 0..n {
            for j in (i + 1)..n {
                if points[i].distance_squared(points[j]) < threshold_sq {
                    edge_count += 1;
                }
            }
        }

        // From Euler's formula: V - E + F = 2 for connected planar graph
        // β₁ = E - V + 1 for a connected component
        // This is a rough estimate
        let components = self.count_components(points);
        if edge_count > n + components {
            edge_count - n - components + 1
        } else {
            0
        }
    }

    /// Estimate voids (simplified)
    fn estimate_voids(&self, points: &[Vec4]) -> usize {
        // This is a very rough estimate
        // Actual computation requires computing 2-simplices (triangles) and 3-simplices
        let n = points.len();
        if n < 4 {
            return 0;
        }

        // Check for approximately hollow structures
        let centroid = points.iter().fold(Vec4::zero(), |acc, p| acc + *p)
            * (1.0 / n as f64);

        let avg_dist: f64 = points.iter().map(|p| p.distance(centroid)).sum::<f64>() / n as f64;

        let dist_variance: f64 = points.iter()
            .map(|p| (p.distance(centroid) - avg_dist).powi(2))
            .sum::<f64>() / n as f64;

        // Low variance in distances to centroid suggests shell-like structure (void)
        if dist_variance < 0.1 * avg_dist * avg_dist && n >= 8 {
            1
        } else {
            0
        }
    }

    /// Detect significant topological signals (changes)
    pub fn detect_signals(&self, points: &[Vec4]) -> Vec<TopologicalSignal> {
        let mut signals = Vec::new();

        // Compare to last computed Betti numbers
        let current = self.compute_betti_without_update(points);

        // Detect changes
        if current.b0 > self.last_betti.b0 {
            signals.push(TopologicalSignal {
                signal_type: SignalType::ClusterBirth,
                persistence: 1.0,
                location: self.estimate_change_location(points, SignalType::ClusterBirth),
                scale: self.distance_threshold,
            });
        } else if current.b0 < self.last_betti.b0 {
            signals.push(TopologicalSignal {
                signal_type: SignalType::ClusterMerge,
                persistence: 1.0,
                location: self.estimate_change_location(points, SignalType::ClusterMerge),
                scale: self.distance_threshold,
            });
        }

        if current.b1 > self.last_betti.b1 {
            signals.push(TopologicalSignal {
                signal_type: SignalType::LoopFormation,
                persistence: 1.0,
                location: self.estimate_change_location(points, SignalType::LoopFormation),
                scale: self.distance_threshold,
            });
        } else if current.b1 < self.last_betti.b1 {
            signals.push(TopologicalSignal {
                signal_type: SignalType::LoopFilled,
                persistence: 1.0,
                location: self.estimate_change_location(points, SignalType::LoopFilled),
                scale: self.distance_threshold,
            });
        }

        if current.b2 > self.last_betti.b2 {
            signals.push(TopologicalSignal {
                signal_type: SignalType::VoidFormation,
                persistence: 1.0,
                location: self.estimate_change_location(points, SignalType::VoidFormation),
                scale: self.distance_threshold,
            });
        }

        signals
    }

    /// Compute Betti without updating last state
    fn compute_betti_without_update(&self, points: &[Vec4]) -> BettiNumbers {
        let b0 = self.count_components(points);
        let b1 = self.estimate_loops(points);
        let b2 = self.estimate_voids(points);
        BettiNumbers { b0, b1, b2, b3: 0 }
    }

    /// Estimate location of topological change
    fn estimate_change_location(&self, points: &[Vec4], _signal_type: SignalType) -> [f64; 4] {
        // Return centroid as approximation
        if points.is_empty() {
            return [0.0; 4];
        }

        let centroid = points.iter().fold(Vec4::zero(), |acc, p| acc + *p)
            * (1.0 / points.len() as f64);

        centroid.to_array()
    }

    /// Get last computed Betti numbers
    pub fn last_betti(&self) -> BettiNumbers {
        self.last_betti
    }

    /// Set distance threshold
    pub fn set_threshold(&mut self, threshold: f64) {
        self.distance_threshold = threshold;
    }
}

impl Default for HomologyAnalyzer {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_betti_numbers_basic() {
        let betti = BettiNumbers { b0: 2, b1: 1, b2: 0, b3: 0 };
        assert_eq!(betti.total(), 3);
        assert_eq!(betti.euler_characteristic(), 1);
    }

    #[test]
    fn test_single_point() {
        let mut analyzer = HomologyAnalyzer::new();
        let points = vec![Vec4::new(0.0, 0.0, 0.0, 0.0)];
        let betti = analyzer.compute_betti(&points);

        assert_eq!(betti.b0, 1); // Single component
        assert_eq!(betti.b1, 0); // No loops
    }

    #[test]
    fn test_two_distant_points() {
        let mut analyzer = HomologyAnalyzer::new();
        analyzer.set_threshold(1.0);

        let points = vec![
            Vec4::new(0.0, 0.0, 0.0, 0.0),
            Vec4::new(10.0, 0.0, 0.0, 0.0), // Far apart
        ];
        let betti = analyzer.compute_betti(&points);

        assert_eq!(betti.b0, 2); // Two separate components
    }

    #[test]
    fn test_connected_points() {
        let mut analyzer = HomologyAnalyzer::new();
        analyzer.set_threshold(2.0);

        let points = vec![
            Vec4::new(0.0, 0.0, 0.0, 0.0),
            Vec4::new(1.0, 0.0, 0.0, 0.0),
            Vec4::new(2.0, 0.0, 0.0, 0.0),
        ];
        let betti = analyzer.compute_betti(&points);

        assert_eq!(betti.b0, 1); // Single connected component
    }
}
